#' @title FoodSeq Setup Pipeline
#'
#' @description This function runs raw phyloseq object through typical inital processing steps.
#'
#' @param physeq raw phyloseq object
#' @param amplicon trnl or 12s
#' @param sepVar define sam_data variable by which CLR transform will be done separately (e.g., seq_date, Study)
#' @param collapse optional to run collapseNoMismatch()
#' @param CommonNames optional space to add common names file
#' @param CommonNameCol optional to add name of common name column to make a new column that's labeled with common name if available and lowestLevel if not
#' @param oldplantNames toggle on/off when using 1:1 asv matching for plant common names -- for using old common names file (not generated by Ashish) -- this parameter is temporary and will be phased out after Ashish adjusts common_name column to have shorter length.
#'
#' @return ps.raw = phyloseq object with no changes from physeq other than adding p/aMR, Shannon, and reads to samdf (no NA/human reads removed)
#' @return ps = phyloseq object with no changes from physeq other than changes to ps.raw + animals glommed + NA/human reads removed
#' @return ps.ra = changes to ps (but NA's conserved) + transformed to relative abundance
#' @return ps.filt.clr = changes to ps + clr transformation + NA/human reads removed
#' @return ps.na = phyloseq object sub-setted for samples containing NA reads
#'
#'
#' @importFrom dada2 collapseNoMismatch
#' @importFrom vegan diversity
#' @importFrom vegan specnumber
#'
#' @export
foodseqSetup <- function(physeq,
                         amplicon = "",
                         sepVar = NULL,
                         collapse = FALSE,
                         CommonNames = NULL,
                         CommonNameCol = NULL,
                         oldplantNames = TRUE){
  ps <- physeq
  amplicon <- tolower(amplicon) # change casing for matching

  # ensure that colnames(tax_table) are all lowercase
  taxdf <- ps@tax_table %>%
    data.frame()

  colnames(taxdf) <- tolower(colnames(taxdf))

  tax_table(ps) <- taxdf %>%
    as.matrix() %>%
    tax_table()

  # Initializes sample data if you're working with raw phyloseq
  if(is.null(ps@sam_data)) {
    ids <- sample_names(ps)
    samdf <- data.frame(row.names = ids) %>%
      mutate(Sample_ID = row.names(.))
    sample_data(ps) <- sample_data(samdf)
  }

  ##############################################################################
  # trnL
  if(amplicon == "trnl") {

    if(collapse) {
      # CollapseNoMismatch
      seqtab.merged <- ps@otu_table

      before_collapse <- dim(seqtab.merged) # Check dimensions before and after collapse
      print(paste("Before collapse-- ", "Rows: ", before_collapse[1], "Columns: ", before_collapse[2]))

      seqtab.merged <- dada2::collapseNoMismatch(seqtab.merged)

      after_collapse <- dim(seqtab.merged)
      print(paste("After collapse-- ", "Rows: ", after_collapse[1], "Columns: ", after_collapse[2]))
      print(paste("Taxa Collapsed: ", before_collapse[2] - after_collapse[2]))

      otu_table(ps) <- otu_table(seqtab.merged, taxa_are_rows = FALSE) # Add back OTU table to ps
    }

    # Add lowestLevel
    tax_table(ps) <- ps@tax_table %>%
      data.frame() %>%
      select(c(varietas, forma, subspecies, species, genus, family, order, class, phylum, superkingdom)) %>% # remove potential duplicates
      mutate(lowestLevel = coalesce(varietas, forma, species, genus, family, order, class, phylum, superkingdom)) %>%
      as.matrix() %>%
      tax_table()

    # Add common names
    if(!is.null(CommonNames)) {
      if (oldplantNames) { # This is based on the old common names file that is a 1-to-1 match for ASV's
        taxcols <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species", "subspecies", "forma", "varietas", "lowestLevel")

        tax_table(ps) <- ps@tax_table %>%
          data.frame() %>%
          rownames_to_column(var = "asv") %>%
          select(asv, any_of(taxcols)) %>% # prevents double assignment of column names
          left_join(CommonNames, by = "asv") %>%
          column_to_rownames(var = "asv") %>%
          as.matrix() %>%
          tax_table()

        # Add taxonomy label that defaults to common name but fills in with lowestLevel if NA
        if(!is.null(CommonNameCol)) {
          tax_table(ps) <- ps@tax_table %>%
            data.frame() %>%
            mutate(taxlabel = ifelse(is.na(.data[[CommonNameCol]]) | trimws(.data[[CommonNameCol]]) == "", lowestLevel, .data[[CommonNameCol]])) %>%
            as.matrix() %>%
            tax_table()
        }
      } else { # This is based on Ashish's new common names file
        cols <- c("asv", "species", "genus", "family", "order", "class", "phylum", "superkingdom", "lowestLevel")

        taxtab <- ps@tax_table %>%
          data.frame() %>%
          rownames_to_column(var = "asv") %>%
          dplyr::select(all_of(cols))

        taxtab <- taxtab %>%
          mutate(
            scientific_name = trimws(coalesce(species, genus, family, order, class, phylum, superkingdom))  # Choose the lowest assigned level; for trnL add Varietas and Forma
          )

        plantnames <- CommonNames %>%
          dplyr::rename(scientific_name = name) %>%
          dplyr::select(scientific_name, common_name)

        df <- left_join(taxtab, plantnames, by = join_by(scientific_name))

        result <- df %>%
          group_by(asv) %>%
          summarize(
            # Find the most specific common taxonomic classification
            name = {
              ranks <- c("species", "genus", "family", "order", "class", "phylum", "superkingdom")  # Specific to general
              common_rank <- ranks[sapply(ranks, function(rank) {
                # Exclude NA and check if all remaining values are identical
                values <- na.omit(cur_data()[[rank]])
                length(unique(values)) == 1
              })][1]

              if (!is.null(common_rank) && !is.na(common_rank)) {
                # Return the single shared value for the rank
                unique(na.omit(cur_data()[[common_rank]]))
              } else {
                NA_character_  # If no common rank is found, return NA
              }
            },
            # Concatenate scientific names
            taxon = paste(unique(scientific_name), collapse = "; "),
            # Concatenate common names
            common_name = paste(na.omit(unique(common_name)), collapse = "; ")
          ) %>%
          ungroup()  %>% # Remove grouping
          select(asv, common_name)

        tax_table(ps) <- ps@tax_table %>%
          data.frame() %>%
          rownames_to_column(var = "asv") %>%
          select(all_of(cols)) %>%
          left_join(result, by = "asv") %>%
          column_to_rownames(var = "asv") %>%
          as.matrix() %>%
          tax_table()

        # Add taxonomy label that defaults to common name but fills in with lowestLevel if NA
        if(!is.null(CommonNameCol)) {
          tax_table(ps) <- ps@tax_table %>%
            data.frame() %>%
            mutate(taxlabel = ifelse(is.na(.data[[CommonNameCol]]) | trimws(.data[[CommonNameCol]]) == "", lowestLevel, .data[[CommonNameCol]])) %>%
            as.matrix() %>%
            tax_table()
        }
      }
    }

    # Relative Abundance
    ps.ra <- transform_sample_counts(ps, function(x){x/sum(x)})

    # CLR-transform and filter
    if (!is.null(sepVar)) {
      num <- ps@sam_data[[sepVar]] %>%
        unique()

      clr.list <- list()

      # Perform filter within sepVar
      for (i in num) {
        ps.temp <- prune_samples(get_variable(ps, sepVar) %in% i, ps) %>%
          prune_taxa(taxa_sums(.) > 0, .)

        ps.filt.clr <- ps.temp %>%
          prune_samples(sample_sums(.) >0, .) %>% # Remove samples that do not have any reads, will mess up PCA plot
          prune_taxa(taxa_sums(.) > 0, .) %>% # Remove taxa that have zero reads across included samples
          microbiome::transform(transform = "compositional") %>%
          microbiome::transform(transform = "clr") %>%
          subset_taxa(!is.na(superkingdom)) # Remove non-foods

        clr.list[[i]] <- ps.filt.clr
      }

      # Merge separate CLR filtered phyloseqs
      ps.filt.clr <- do.call(merge_phyloseq, clr.list)

    } else { # Perform CLR filter within the entire phyloseq object
      ps.filt.clr <- ps %>%
        prune_samples(sample_sums(.) >0, .) %>% # Remove samples that do not have any reads, will mess up PCA plot
        prune_taxa(taxa_sums(.) > 0, .) %>% # Remove taxa that have zero reads across included samples
        microbiome::transform(transform = "compositional") %>%
        microbiome::transform(transform = "clr") %>%
        subset_taxa(!is.na(superkingdom)) # Remove non-foods
    }

    # Update read counts in phyloseq object
    sample_data(ps)$reads <- sample_sums(ps)

    ps.filt.clr@sam_data$reads <- 0 # initialize reads variable in ps.filt.clr
    matching_samples <- intersect(rownames(ps.filt.clr@sam_data), rownames(ps@sam_data))
    sample_data(ps.filt.clr)[matching_samples, "reads"] <- sample_data(ps)[matching_samples, "reads"]

    # pMR
    pMR <- ifelse(subset_taxa(ps.filt.clr, !is.na(superkingdom))@otu_table>0,1,0) %>%      # Converts OTU table values to presence/absence
      rowSums()
    ps.filt.clr@sam_data$pMR <- pMR

    ps@sam_data$pMR <- 0 # initialize pMR variable in ps
    matching_samples <- rownames(ps.filt.clr@sam_data) %in% rownames(ps@sam_data)
    ps@sam_data[rownames(ps.filt.clr@sam_data)[matching_samples], "pMR"] <- ps.filt.clr@sam_data$pMR

    # Shannon Index
    shannon <- estimate_richness(ps, split = TRUE, measures = "Shannon") %>%
      rownames_to_column(var = "samid")
    ps@sam_data$Shannon_diversity_plants <- shannon$Shannon

    sample_data(ps.filt.clr) <- ps.filt.clr@sam_data %>%
      data.frame() %>%
      rownames_to_column(var = "samid") %>%
      left_join(shannon, by = "samid") %>%
      mutate(Shannon = ifelse(is.na(Shannon), 0, Shannon)) %>%
      select(-any_of("Shannon_diversity_plants")) %>%
      dplyr::rename(Shannon_diversity_plants = Shannon) %>%
      column_to_rownames(var = "samid") %>%
      sample_data()

    # Simpson's Evenness
    simpson_reciprocal <- vegan::diversity(otu_table(ps.ra), index = "invsimpson")
    richness <- vegan::specnumber(otu_table(ps))
    simpson_evenness <- (simpson_reciprocal / richness) %>%
      data.frame(Simpson_evenness_plants = .) %>%
      rownames_to_column(var = "samid")

    ps@sam_data$Simpson_evenness_plants <- simpson_evenness$Simpson_evenness_plants
    sample_data(ps.filt.clr) <- ps.filt.clr@sam_data %>%
      data.frame() %>%
      rownames_to_column(var = "samid") %>%
      select(-any_of("Simpson_evenness_plants")) %>%
      left_join(simpson_evenness, by = "samid") %>%
      mutate(Simpson_evenness_plants = ifelse(is.na(Simpson_evenness_plants), 0, Simpson_evenness_plants)) %>%
      column_to_rownames(var = "samid") %>%
      sample_data()
    sample_data(ps.ra) <- ps.ra@sam_data %>%
      data.frame() %>%
      rownames_to_column(var = "samid") %>%
      select(-any_of("Simpson_evenness_plants")) %>%
      left_join(simpson_evenness, by = "samid") %>%
      mutate(Simpson_evenness_plants = ifelse(is.na(Simpson_evenness_plants), 0, Simpson_evenness_plants)) %>%
      column_to_rownames(var = "samid") %>%
      sample_data()



    # ps.na
    ps.na <- subset_taxa(ps, is.na(superkingdom))
  }

  ##############################################################################
  # 12SV5
  if(amplicon == "12s") {
    # Replace "NA" strings with NA value
    tax_table(ps) <- tax_table(ps) %>%
      as.data.frame() %>%
      mutate(across(everything(), ~ ifelse(. == "NA", NA, .))) %>%
      as.matrix() %>%
      tax_table()

    # Add lowestLevel
    tax_table(ps) <- ps@tax_table %>%
      data.frame() %>%
      select(c(subspecies, species, genus, family, order, class, phylum, kingdom)) %>% # remove potential duplicates
      mutate(lowestLevel = coalesce(subspecies, species, genus, family, order, class, phylum, kingdom)) %>%
      as.matrix() %>%
      tax_table()

    # Add common names -- this is based on Ashish's new 12SV5 common names file
    if (!is.null(CommonNames)) {
      cols <- c("asv", "species", "genus", "family", "order", "class", "phylum", "kingdom", "lowestLevel")

      taxtab <- ps@tax_table %>%
        data.frame() %>%
        rownames_to_column(var = "asv") %>%
        dplyr::select(all_of(cols))

      taxtab <- taxtab %>%
        mutate(
          scientific_name = trimws(coalesce(species, genus, family, order, class, phylum, kingdom))  # Choose the lowest assigned level; for trnL add Varietas and Forma
        )

      animalnames <- CommonNames %>%
        dplyr::rename(scientific_name = name) %>%
        dplyr::select(scientific_name, common_name)

      df <- left_join(taxtab, animalnames, by = join_by(scientific_name))

      result <- df %>%
        group_by(asv) %>%
        summarize(
          # Find the most specific common taxonomic classification
          name = {
            ranks <- c("species", "genus", "family", "order", "class", "phylum", "kingdom")  # Specific to general
            common_rank <- ranks[sapply(ranks, function(rank) {
              # Exclude NA and check if all remaining values are identical
              values <- na.omit(cur_data()[[rank]])
              length(unique(values)) == 1
            })][1]

            if (!is.null(common_rank) && !is.na(common_rank)) {
              # Return the single shared value for the rank
              unique(na.omit(cur_data()[[common_rank]]))
            } else {
              NA_character_  # If no common rank is found, return NA
            }
          },
          # Concatenate scientific names
          taxon = paste(unique(scientific_name), collapse = "; "),
          # Concatenate common names
          common_name = paste(na.omit(unique(common_name)), collapse = "; ")
        ) %>%
        ungroup()  %>% # Remove grouping
        select(asv, common_name)

      tax_table(ps) <- ps@tax_table %>%
        data.frame() %>%
        rownames_to_column(var = "asv") %>%
        select(all_of(cols)) %>%
        left_join(result, by = "asv") %>%
        column_to_rownames(var = "asv") %>%
        relocate(lowestLevel, .after = everything()) %>%
        as.matrix() %>%
        tax_table()

      # Add taxonomy label that defaults to common name but fills in with lowestLevel if NA
      if(!is.null(CommonNameCol)) {
        tax_table(ps) <- ps@tax_table %>%
          data.frame() %>%
          mutate(taxlabel = ifelse(is.na(.data[[CommonNameCol]]) | trimws(.data[[CommonNameCol]]) == "", lowestLevel, .data[[CommonNameCol]])) %>%
          relocate(lowestLevel, .after = everything()) %>%
          as.matrix() %>%
          tax_table()
      }
    }

    # Add human_percent to sam_data for each individual
    human_asvs <- ps@tax_table %>%
      data.frame() %>%
      rownames_to_column(var = "asv") %>%
      dplyr::filter(species == "Homo sapiens") %>%
      pull(asv)

    if(length(human_asvs) == 0){
      warning(paste0("\n", "\n", "This phyloseq object has 0 human reads. This is highly suspicious.", "\n",
                     "Results of any subsequent CLR trasnformation results are unreliable.", "\n",
                     "All reads (human/non-foods and NA) are neeed for proper transformation", "\n",
                     "I recommend using the full phyloseq object.", "\n", "\n"))

    } else {
      total_reads <- ps@otu_table %>%
        data.frame() %>%
        rownames_to_column(var = "sample") %>%
        pivot_longer(-sample, names_to = "asv") %>%
        mutate(value = as.numeric(value)) %>% # ensure numeric
        group_by(sample) %>%
        summarise(total_reads = sum(value, na.rm = TRUE))

      human_reads <- ps@otu_table %>%
        data.frame() %>%
        rownames_to_column(var = "sample") %>%
        pivot_longer(-sample, names_to = "asv") %>%
        mutate(value = as.numeric(value)) %>% # ensure numeric
        dplyr::filter(asv %in% human_asvs) %>%
        group_by(sample) %>%
        summarise(human_reads = sum(value, na.rm = TRUE))

      human_percent <- total_reads %>%
        left_join(human_reads, by = "sample") %>%
        mutate(human_reads_percent = human_reads/total_reads*100) %>%
        select(sample, human_reads, human_reads_percent)

      sample_data(ps) <- ps@sam_data %>%
        data.frame() %>%
        rownames_to_column(var = "sample") %>%
        left_join(human_percent, by = "sample") %>%
        column_to_rownames(var = "sample") %>%
        sample_data()

      # Calculate and print % human reads
      total_human_reads <- ps %>%
        subset_taxa(species == "Homo sapiens") %>%
        sample_sums()

      percent_human_reads <- round((sum(total_human_reads)/sum(total_reads$total_reads))*100, 2)

      print(paste0(percent_human_reads, "% reads were assigned to Homo sapien"))
    }

    total_reads <- sample_sums(ps)

    # Glom animals
    ps.glom <- tax_glom(ps, taxrank = "lowestLevel")

    # Relative Abundance
    ps.ra <- transform_sample_counts(ps.glom, function(x){x/sum(x)})

    # CLR-transform and Filter NA's
    if (!is.null(sepVar)) {
      num <- ps@sam_data[[sepVar]] %>%
        unique()

      clr.list <- list()

      # Perform separated CLR-filtering
      for (i in num) {
        ps.temp <- prune_samples(get_variable(ps.glom, sepVar) %in% i, ps.glom) %>%
          prune_taxa(taxa_sums(.) > 0, .)

        ps.filt.clr <- ps.temp %>%
          prune_samples(sample_sums(.) >0, .) %>% # Remove samples that do not have any reads, will mess up PCA plot
          prune_taxa(taxa_sums(.) > 0, .) %>% # Remove taxa that have zero reads across included samples
          microbiome::transform(transform = "compositional") %>%
          microbiome::transform(transform = "clr") %>%
          subset_taxa(!is.na(kingdom)) %>%  # Remove non-foods
          subset_taxa(lowestLevel != "Homo sapiens") # Remove human reads

        clr.list[[i]] <- ps.filt.clr
      }

      # Merge separate CLR filtered phyloseqs
      ps.filt.clr <- do.call(merge_phyloseq, clr.list)
    } else { # Perform regular CLR-filtering within entire phyloseq object
      ps.filt.clr <- ps.glom %>%
        prune_samples(sample_sums(.) >0, .) %>% # Remove samples that do not have any reads, will mess up PCA plot
        prune_taxa(taxa_sums(.) > 0, .) %>% # Remove taxa that have zero reads across included samples
        microbiome::transform(transform = "compositional") %>%
        microbiome::transform(transform = "clr") %>%
        subset_taxa(!is.na(kingdom)) %>%  # Remove non-foods
        subset_taxa(lowestLevel != "Homo sapiens") # Remove human reads
    }

    # Remove human reads
    ps.noHuman <- ps.glom %>%
      subset_taxa(lowestLevel != "Homo sapiens")

    # aMR
    aMR <- ifelse(subset_taxa(ps.noHuman, !is.na(kingdom))@otu_table>0,1,0) %>%      # Converts OTU table values to presence/absence
      rowSums()
    ps.noHuman@sam_data$aMR <- aMR

    ps@sam_data$aMR <- 0 # initialize aMR variable in ps
    matching_samples <- rownames(ps.noHuman@sam_data) %in% rownames(ps@sam_data)
    ps@sam_data[rownames(ps.noHuman@sam_data)[matching_samples], "aMR"] <- ps.noHuman@sam_data$aMR[matching_samples]

    ps.filt.clr@sam_data$aMR <- 0 # initialize aMR variable in ps.filt.clr
    matching_samples <- rownames(ps.noHuman@sam_data) %in% rownames(ps.filt.clr@sam_data)
    ps.filt.clr@sam_data[rownames(ps.noHuman@sam_data)[matching_samples], "aMR"] <- ps.noHuman@sam_data$aMR[matching_samples]

    ps.ra@sam_data$aMR <- 0 # initialize aMR variable in ps.ra
    matching_samples <- rownames(ps.noHuman@sam_data) %in% rownames(ps.ra@sam_data)
    ps.ra@sam_data[rownames(ps.noHuman@sam_data)[matching_samples], "aMR"] <- ps.noHuman@sam_data$aMR

    physeq@sam_data$aMR <- 0 # initialize aMR variable in physeq
    matching_samples <- rownames(ps.noHuman@sam_data) %in% rownames(physeq@sam_data)
    physeq@sam_data[rownames(ps.noHuman@sam_data)[matching_samples], "aMR"] <- ps.noHuman@sam_data$aMR

    # Shannon Index
    shannon <- estimate_richness(ps.noHuman, split = TRUE, measures = "Shannon")
    ps.noHuman@sam_data$Shannon_diversity_animals <- shannon$Shannon

    # Update read counts
    sample_data(ps)$reads <- sample_sums(ps)

    ps.filt.clr@sam_data$reads <- 0 # initialize reads variable in ps.filt.clr
    matching_samples <- intersect(rownames(ps.filt.clr@sam_data), rownames(ps@sam_data))
    sample_data(ps.filt.clr)[matching_samples, "reads"] <- sample_data(ps)[matching_samples, "reads"]

    # ps.na
    ps.na <- subset_taxa(ps, is.na(kingdom))
  }

  # ### Beta diversity metrics -- This needs to be incorporated later
  # bc_dist <- phyloseq::distance(ps.ra, method = "bray") # non-clr filtered
  # bc_matrix <- as.matrix(bc_dist)

  return(list(ps.raw = physeq, # Raw ps with no changes, but p/aMR, Shannon, and reads are updated (no NA/human filtered)
              ps = ps, # Animals glommed + NA/human reads removed
              ps.ra = ps.ra, # Relative abundance, includes NA's
              ps.filt.clr = ps.filt.clr, # Foods only, CLR-transformed
              ps.na = ps.na # Raw ps with only is.na(superkingdom) samples
              #bc_dist = bc_dist
              #bc_matrix = bc_matrix
  ))
}
